pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/IngestionDelayTracker.java

```diff
@@ -0,0 +1,361 @@
+    }
+    // If we are consuming we do not need to track this partition for removal.
+    _partitionsMarkedForVerification.remove(partitionGroupId);
+  }
+
+  /*
+   * Handle partition removal event. This must be invoked when we stop serving a given partition for
+   * this table in the current server.
+   *
+   * @param partitionGroupId partition id that we should stop tracking.
+   */
+  public void stopTrackingPartitionIngestionDelay(int partitionGroupId) {
+    removePartitionId(partitionGroupId);
+  }
+
+  /*
+   * This method is used for timing out inactive partitions, so we don't display their metrics on current server.
+   * When the inactive time exceeds some threshold, we read from ideal state to confirm we still host the partition,
+   * if not we remove the partition from being tracked locally.
+   * This call is to be invoked by a timer thread that will periodically wake up and invoke this function.
+   */
+  public void timeoutInactivePartitions() {
+    Set<Integer> partitionsHostedByThisServer = null;
+    // Check if we have any partition to verify, else don't make the call to check ideal state as that
+    // involves network traffic and may be inefficient.
+    ArrayList<Integer> partitionsToVerify = getPartitionsToBeVerified();
+    if (partitionsToVerify.size() == 0) {
+      // Don't make the call to getHostedPartitionsGroupIds() as it involves checking ideal state.
+      return;
+    }
+    try {
+      partitionsHostedByThisServer = _realTimeTableDataManager.getHostedPartitionsGroupIds();
+    } catch (Exception e) {
+      _logger.error("Failed to get partitions hosted by this server, table={}", _tableNameWithType);
+      return;
+    }
+    for (int partitionGroupId : partitionsToVerify) {
+      if (!partitionsHostedByThisServer.contains(partitionGroupId)) {
+        // Partition is not hosted in this server anymore, stop tracking it
+        removePartitionId(partitionGroupId);
+      }
+    }
+  }
+
+  /*
+   * This function is invoked when a partition goes from CONSUMING to ONLINE, so we can assert whether the
+   * partition is still hosted by this server after some interval of time.
+   *
+   * @param partitionGroupId Partition id that we need confirmed via ideal state as still hosted by this server.
+   */
+  public void markPartitionForVerification(int partitionGroupId) {
+    _partitionsMarkedForVerification.put(partitionGroupId, _clock.millis());
+  }
+
+  /*
+   * This is the function to be invoked when reading the metric.
+   * It reports the maximum ingestion delay for all partitions of this table being served
+   * by current server; it adds the time elapsed since the sample was taken to the measure.
+   * If no measures have been taken, then the reported value is zero.
+   *
+   * @return max ingestion delay in milliseconds.
+   */
+  public long getMaximumIngestionDelay() {
+    DelayMeasure currentMaxDelay = getMaximumDelay();
```

Comment:
I don't think this is correct because when calculating the `currentMaxDelay`, we didn't count for the difference between current time and sample time. E.g. we might get a message with very small delay initially, but the stream hangs for a long time

Reply:
I see your point, as suggested I will remove this to start with

Reply:
removed aggregate metric, see commit 3ce7b783c5dc88073106f400f00783ba4ab71281